File: Week06/LeetCode-0347-3.py
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        """Here is an O(n) solution"""
        myMap = dict()
        
        # Count the occurrence of each number
        for num in nums:
            if num not in myMap: myMap[num] = 1
            else: myMap[num] += 1

        # Find the freq. of the number that occurs the most
        maxCount = 0
        for freq in myMap.values():
            maxCount = max(maxCount, freq)

        # Create a LUT with the most freq. numbers        
        freqList = []
        for i in range(maxCount+1): freqList.append([])
        for num, freq in myMap.items():
            freqList[freq].append(num)

        # Copy the numbers from the freq. list to the result
        result = []
        index = maxCount
        while k > 0 and index >= 0:
            for i in range(len(freqList[index])):
                if k == 0: break
                result.append(freqList[index][i])                    
                k -= 1
            index -= 1
        return result
        

File: Week06/LeetCode-0347-3.cpp
class Solution {
public:
    ///-------------------------------------------------------------
    /// Here is an O(n) solution
    ///
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> map;

        // Count the occurrence of each number
        for (auto num: nums) map[num]++;

        // Find the freq. of the number that occurs the most
        int maxCount = 0;
        for (auto p: map) if (p.second > maxCount) maxCount = p.second;

        // Create a LUT with the most freq. numbers (assumes INT_MINT is not a valid number)
        vector<vector<int>> freq(maxCount+1);
        for (auto [f, num]: map) freq[f].push_back(num);

        // Copy the numbers from the freq. list to the result
        vector<int> result;
        for (int index=maxCount; k>0 && index >= 0; index--){
            for (int i=0; i<freq[index].size() && k>0; i++, k--){
                result.push_back(freq[index][i]);
            } //end-for
        } //end-while

        return result;
    } //end-topKFrequent
};
